peer/peer.go

```diff
@@ -1,174 +1,268 @@
-type Peer struct {
-	host.Host
-}
-
-// New creates a Cumulus host with the given IP addr and TCP port.
-// This may throw an error if we fail to create a key pair, a pid, or a new
-// multiaddress.
-func New(ip string, port int) (*Peer, error) {
-	// Generate a key pair for this host. We will only use the pudlic key to
-	// obtain a valid host ID.
-	// Cannot throw error with given arguments
-	priv, pub, _ := crypto.GenerateKeyPair(crypto.RSA, 2048)
+// PStore stores information about every peer we are connected to. All peers we
+// connect to should have a reference to this peerstore so they can populate it.
+var PStore = &PeerStore{peers: make(map[string]*Peer, 0)}
 
-	// Obtain Peer ID from public key.
-	// Cannot throw error with given argument
-	pid, _ := lpeer.IDFromPublicKey(pub)
+// PeerStore is a thread-safe container for all the peers we are currently
+// connected to.
+type PeerStore struct {
+	peers map[string]*Peer
+	lock  sync.RWMutex
+}
 
-	// Create a multiaddress (IP address and TCP port for this peer).
-	addr, err := ma.NewMultiaddr(fmt.Sprintf("/ip4/%s/tcp/%d", ip, port))
-	if err != nil {
-		return nil, err
+// NewPeerStore returns an initialized peerstore.
+func NewPeerStore() *PeerStore {
+	return &PeerStore{
+		peers: make(map[string]*Peer, 0),
+		lock:  sync.RWMutex{},
 	}
+}
 
-	// Create Peerstore and add host's keys to it (avoids annoying err)
-	ps := pstore.NewPeerstore()
-	ps.AddPubKey(pid, pub)
-	ps.AddPrivKey(pid, priv)
-
-	// Create swarm (this is the interface to the libP2P Network) using the
-	// multiaddress, peerID, and peerStore we just created.
-	netwrk, err := swarm.NewNetwork(
-		context.Background(),
-		[]ma.Multiaddr{addr},
-		pid,
-		ps,
-		nil)
-	if err != nil {
-		return nil, err
-	}
+// Add synchronously adds the given peer to the peerstore
+func (ps *PeerStore) Add(p *Peer) {
+	ps.lock.Lock()
+	defer ps.lock.Unlock()
+	ps.peers[p.ID.String()] = p
+}
 
-	// Actually create the host and peer with the network we just set up.
-	host := bhost.New(netwrk)
-	peer := &Peer{Host: host}
+// Remove synchronously removes the given peer from the peerstore
+func (ps *PeerStore) Remove(id string) {
+	ps.lock.Lock()
+	defer ps.lock.Unlock()
+	delete(ps.peers, id)
+}
 
-	// Build host multiaddress
-	hostAddr, _ := ma.NewMultiaddr(fmt.Sprintf("/ipfs/%s",
-		host.ID().Pretty()))
+// Get synchronously retreives the peer with the given id from the peerstore
+func (ps *PeerStore) Get(id string) *Peer {
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+	return ps.peers[id]
+}
 
-	// Now we can build a full multiaddress to reach this host
-	// by encapsulating both addresses:
-	fullAddr := addr.Encapsulate(hostAddr)
-	log.Info("I am ", fullAddr)
+// Addrs returns the list of addresses of the peers in the peerstore in the form
+// <IP addr>:<port>
+func (ps *PeerStore) Addrs() []string {
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+	addrs := make([]string, len(ps.peers), len(ps.peers))
```

Comment:
Irrelevant now that I switched the peerstore back to a map. Didn't know we could iterate over a map back then.... my bad.